packages/sources/coinmetrics-test/src/endpoint/price-ws.ts

```diff
@@ -0,0 +1,194 @@
+      context.adapterConfig.WS_API_ENDPOINT || DEFAULT_WS_API_ENDPOINT,
+    ).toString()
+    logger.debug(`Generated the following url: ${generated}`)
+    return generated
+  },
+  handlers: {
+    open(connection) {
+      return new Promise((resolve) => {
+        // Set up listener
+        logger.debug('Opening websocket connection')
+        logger.debug(connection, 'Connection')
+
+        connection.on('message', (data: WebSocketRawData) => {
+          const parsed = JSON.parse(data.toString())
+          logger.debug(parsed, 'received message on connection')
+          resolve()
+        })
+      })
+    },
+
+    message(message: WsAssetMetricsMessage): ProviderResult<WsEndpointTypes>[] | undefined {
+      logger.trace(message, 'Got response from websocket')
+
+      //Cast message to all of its known types so we can check which variant we received
+      // const successMessage = message as WsAssetMetricsSuccessResponse
+      // const reorgMessage = message as WsAssetMetricsSuccessResponse
+      const successMessage = message as WsAssetMetricsSuccessResponse
+      const errorMessage = message as WsAssetMetricsErrorResponse
+      if (successMessage.asset) {
+        // Success message
+        let metric: number | undefined
+        let metricQuote = ''
+        for (const quote of VALID_REFERENCE_RATE_QUOTES) {
+          metric = successMessage[`ReferenceRate${quote}`]
+          if (metric) {
+            metricQuote = quote
+            break
+          }
+        }
+        if (!metric) {
+          logger.debug(successMessage, `No 'ReferenceRate($QUOTE) metric found in message`)
+        }
+        return [
+          {
+            params: {
+              base: successMessage.asset,
+              quote: metricQuote,
+              endpoint: 'price-ws',
+            },
+            value: metric || 0, // Fallback zero here just to get past the type checker. We've already checked that metric is defined above.
+          },
+        ]
+      } else if (errorMessage.error) {
+        logger.error(errorMessage, `Error response from websocket`)
+      } else {
+        logger.warn(message, 'Unknown message type from websocket')
+      }
+      return undefined
+    },
+  },
+  builders: {
+    subscribeMessage: (params: WsAssetMetricsRequestBody) => {
+      const key = getSubscriptionKey(params)
+      logger.debug(
+        `Changing currentBase from ${currentBase} to ${params.base.toLowerCase()} and currentQuote from ${currentQuote} to ${params.quote.toUpperCase()}`,
+      )
+      currentBase = params.base.toLowerCase()
+      currentQuote = params.quote.toUpperCase()
+      logger.debug(`Subscribing with this key: ${key}`)
+      return key
+    },
+    unsubscribeMessage: (params: WsAssetMetricsRequestBody) => {
+      const key = getSubscriptionKey(params)
+      logger.debug(`Unsubscribing with this key: ${key}`)
+      return key
+    },
+  },
```

Comment:
Seems like we should make builders optional for this transport, no? In this case they are unnecessary